project/src/ui/help_viewer/HelpViewer.gd

```diff
@@ -0,0 +1,222 @@
+	if "**note:**" in line:
+		line = line.replacen("_**note:**", "note:")
+		line = line.trim_suffix("_")
+		line = "[i]" + line + "[/i]"
+
+	# Image
+	var img_width = 464  # depends on the window width
+	if line.begins_with("![](https://i.imgur.com/"):
+		line = line.replacen("![](", "").replacen(")", "")
+		var image_file_name = line.split("/")[-1]
+		_download_image(line, USER_DIR + image_file_name)
+		line = "[img=<" + str(img_width) + ">]" + USER_DIR + image_file_name + "[/img]"
+
+	# Link
+	var hyperlink_position_start = 0
+	var hyperlink_position_end_space = null
+	var hyperlink_position_end_bracket = null
+	var hyperlink_position_end_backslash = null
+	var hyperlink_position_end = null
+	if "https://" in line:
+		hyperlink_position_start = line.find("https://")
+		hyperlink_position_end_space = line.findn(" ", hyperlink_position_start)
+		hyperlink_position_end_bracket = line.findn(")", hyperlink_position_start)
+		hyperlink_position_end_backslash = line.findn("\\", hyperlink_position_start)
+		if hyperlink_position_end_space == -1:
+			hyperlink_position_end_space = INT64_MAX
+		if hyperlink_position_end_bracket == -1:
+			hyperlink_position_end_bracket = INT64_MAX
+		if hyperlink_position_end_backslash == -1:
+			hyperlink_position_end_backslash = INT64_MAX
+		if (
+			hyperlink_position_end_space == INT64_MAX
+			&& hyperlink_position_end_bracket == INT64_MAX
+			&& hyperlink_position_end_backslash == INT64_MAX
+		):
+			hyperlink_position_end = line.length()
+		else:
+			hyperlink_position_end = min(
+				hyperlink_position_end_space,
+				min(hyperlink_position_end_bracket, hyperlink_position_end_backslash)
+			)
+		line = line.insert(hyperlink_position_start, "[url]")
+		line = line.insert(hyperlink_position_end + 5, "[/url]")  # for some reason doesn't end at the actual end, need + 5
+
+	line += "\n"
+	return [line, code_snippet]
+
+
+
+func _richtextlabel_on_meta_clicked(meta):
+	OS.shell_open(str(meta))
+
+
+func _gui_input(event: InputEvent):
+	if event.is_action_pressed("mouse_left"):
+		_close()
+
+
+# Added enter_tree() to initialize the screen
+func _enter_tree() -> void:
+	_open()
+
+
+func _close() -> void:
+	emit_signal("exited")
+	var tween = TempTween.new()
+
+	add_child(tween)
+	tween.interpolate_property(self, "rect_scale:y", 1, 0, 0.3, Tween.TRANS_CUBIC)
+	tween.interpolate_property(self, "modulate:a", 1, 0, 0.15)
+	tween.start()
+
+	yield(tween, "tween_all_completed")
+	queue_free()
+
+
+func _open() -> void:
+	var tween = TempTween.new()
+	add_child(tween)
+	tween.interpolate_property(self, "rect_scale:y", 0, 1, 0.15, Tween.TRANS_CUBIC)
+	tween.interpolate_property(self, "modulate:a", 0, 1, 0.15)
+	tween.start()
+
+
+# Find the selected help and update the text window
+func _select_help() -> void:
+	for n in item_list.get_item_count():
+		if item_list.is_selected(n) == true:
+			rich_text_label.clear()
+			rich_text_label.append_bbcode(wiki_pages[n].content)
+
+
+func _download_image(url: String, file_name: String):
```

Comment:
shoudnt this be awaited? bit of a race condition for the caller if the image isnt downloaded quick enough

Reply:
should be resolved now, but a bit confused if yield() is correctly used here

Reply:
the callee would need to `yield(_download_image, "completed")` otherwise the control flow doesn't stop, then same for the callee _markdown_to_bbcode, basically once you yield youd have to do that all the way to the top.

if you are certain bbcode image doesnt care about it then dont do it.
an easy way to check this is by removing your local cache and then starting the app and looking at the docs, If the image never shows up because its still downloading then you have a problem.

Reply:
considering you already yield in your ready function, you should probably just yield everything

Reply:
yeah that was the issue I had, I knew I probably had to do `yield(_download_image, "completed")`, but then I guess I would still need to to the same thing in every function higher up in the chain, which is several spots. But yeah I think there was not a problem before, I will re-check this and maybe just remove the yield

Reply:
How should I handle yielding several different functions which are nested in each other and are of different type? it says that the return of the functions higher up in the chain are not of type object.. but it really works really well when just using yield() on the node..

Reply:
If you want to learn how yields work message me at a later date but otherwise lets keep it how it is :)